Upgraded the RevealDisguiseOn type enum and also added a ValidTargets

StormWingDelta opened the pull request:
Added  a helpful update to Disguise allowing it to be
told to break under more conditions and also so a player can set a list
of valid targets to futher narrow down what Disguise can be used on in
addition to what it is allowed to target.

StormWingDelta committed:
Upgraded the RevealDisguiseOn type enum and also added a ValidTargets

Added the main bulk of a helpful update to Disguise allowing it to be
told to break under more conditions and also so a player can set a list
of valid targets to futher narrow down what Disguise can be used on in
addition to what it is allowed to target.

StormWingDelta commented:
hmm I understand it breaking on the isDocking trait so I can remove that and the other docking stuff unless someone wants it?

As for the other thing it is breaking on I have no idea.

StormWingDelta committed:
Removed the isDocking bool

Since it wasn't being used

StormWingDelta committed:
Fixed Comments of the file

Hopefully I fixed everything, if anything's still wrong after this comment let me know what it is.

StormWingDelta committed:
Corrected as many of the line ending errors and curly brace spacing errors as I could find

I have no idea where any of the others are, at this point I have no idea what it's looking for if there's any more.

GraionDilach commented:
Let's be logical: A good amount of your new triggers are unneeded. Selfhealing dropping disguise would probably be the worst idea ever. Infiltrator dropping the disguise in the last moment? Again notsure on the point.

All the options would make sense only if there would be a way to temporarily drop the disguise and not permanently like how the current system works. With the current system, only the already-there enum options make sense, the others requires too much micro to remain effective.

StormWingDelta commented:
Ya I noticed I went a tad overboard, I'll remove the illogical ones.  Meantime is there some way to setup VS to use the same checks this github is using?  Could swear I saw it somewhere on the wiki.

StormWingDelta committed:
Removed a large number of illogical & unused flags

Got rid of some of the more illogical flags I had put in as well as the ones not even being used.  Also removed the Harvester Docking code as well since the only place it'd have a use is a Thief Harvester. If anyone ants the Heal Damage Check back let me know.

StormWingDelta committed:
Corrected Line Endings & found all stylecop errors

Installed stylecop to check this time and if memory serves you guys are using the LF line ending.  Hopefully I found all the things it was griping about this time.

StormWingDelta commented:
GraionDilach about the making disguise save its last disguise if told to in the yaml.  Sounds like a good addition to disguise if anyone wants to add it.  Might add a way to make disguise auto redisguise as its last disguise if allowed to later unless someone else beats me to it.

GraionDilach commented on OpenRA.Mods.Cnc/Traits/Disguise.cs:
```diff
@@ -76,13 +78,21 @@ class DisguiseInfo : ITraitInfo
 
-		[Desc("Triggers which cause the actor to drop it's disguise. Possible values: None, Attack, Damaged.")]
+		// Added this for more control over when disguises break
+		[Desc("Events leading to the actor breaking Disguise. Possible values are: Attack, Move, Demolish, and Damaged.")]
```
None is still valid.

StormWingDelta replied:
hmm remove None or somehow make use of it?

StormWingDelta replied:
I've got an idea for None assuming it doesn't do it already. How about it block use of the other flags while there? :)

GraionDilach replied:
None is valid and used - asin having the enum on the value None alone will mean it won't drop the disguise in any event. That's how disguise worked in RA2. None is the sole reason I added this enum back in the PR this was introduced.

StormWingDelta committed:
Added None to the string saying what the valid RevealDisguiseTypes are

Yes totally forgot to put that in the string.  Although I've noticed it goes unused in several locations flags are used. Anyone want None to outright block use of the other flags?

StormWingDelta commented:
I'm tempting the idea of turning Disguise into a ConditionalTrait. Anyone else want this?  So far the main use of doing this has helped make the AutoDisguise Trait I'm working on mostly work letting it grab Disguise's Info trait.  As for other uses it could open the door to making Disguise an upgrade for anyone to upgrade into rather than ready to use by default.

StormWingDelta committed:
Added AutoDisguise & changed Disguise into a ConditionalTrait

2nd step in making it so people can make MIrage Tanks is done.  Let me
know if it needs any changes.
Also very fair warning to make this more I've changed Disguise into a
Conditional Trait, Please let me know of any other changes this trait
needs to be fully functional in that respect.

StormWingDelta committed:
Removed the Unused Actor self

I have no clue what to do with the condition manager and the token.  Should I just remove them outright for now or does someone want to help me make use of them in this class?

StormWingDelta commented:
Unsure what to do with the condition manager and its token in the AutoDisguise class if anyone wants to help me find a way to use them give me an idea or I'll just remove them for now.

StormWingDelta committed:
Made use of the unused conditionManager and token

Not sure if I did this correctly but it'll do for now allowing people to
make use of the condition in the info class.

GraionDilach commented:
That should be in a separate/followup PR.

StormWingDelta commented:
Noticed the hard way, Now need to figure out how to revert a PR or branch to an older commit. >.>

StormWingDelta committed:
Revert "Made use of the unused conditionManager and token"

This reverts commit 5beb6943f15b7b98ee31a003996f472baef99a18.

StormWingDelta committed:
Revert "Removed the Unused Actor self"

This reverts commit 7d3d4f15e9c659a90eb3fd3e32b8700a6fb7e187.

StormWingDelta committed:
Revert "Added AutoDisguise & changed Disguise into a ConditionalTrait"

This reverts commit b5f1f5c3825e756f96bb1a83677f2c5f08f29015.

StormWingDelta commented:
Going to upload the Disguise being changed into a Conditional Trait and the Auto Disguise Trait as their own PR to avoid any headaches. :)

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into SWDFeaturesMain

GraionDilach commented on OpenRA.Mods.Cnc/Traits/Disguise.cs:
```diff
@@ -211,14 +223,58 @@ void HandleDisguise(Player oldEffectiveOwner, bool oldDisguiseSetting)
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Attack))
+			if (info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Attack) || info.RevealDisguiseOn.HasFlag(RevealDisguiseType.Demolish))
```
This is wrong.

Demolish triggers the uncloak directly in it's activity and not through an interface,. so i see where this came from. However it doesn't make this any valid - this will render the Demolish value equal to Attack. Either you replace that call in the Demolish activity with an INotifyDemolish interface with this trait also implementing it either you drop this value.

StormWingDelta replied:
hmm didn't notice that particular interface to use, could swear I looked through the list of interfaces trying to find it.  In any event that solves that mystery so it'll get fixed. :)

GraionDilach commented on OpenRA.Mods.Cnc/Traits/Disguise.cs:
```diff
@@ -211,14 +223,58 @@ void HandleDisguise(Player oldEffectiveOwner, bool oldDisguiseSetting)
 
 		void INotifyDamage.Damaged(Actor self, AttackInfo e)
 		{
+			if (e.Damage.Value == 0)
```
This check isn't needed anymore IMO.

StormWingDelta replied:
It felt rather funny when I saw it in a few other places that used it. I'll remove it. :)

GraionDilach commented on OpenRA.Mods.Cnc/Traits/Disguise.cs:
```diff
@@ -211,14 +223,58 @@ void HandleDisguise(Player oldEffectiveOwner, bool oldDisguiseSetting)
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled())
```
We're in the process of deprecating IsDisabled so I don't think it's a good idea to introduce it in this trait.

StormWingDelta replied:
ok, any ideas what to use for if anything? Or just remove it outright?

StormWingDelta committed:
Removed Unneeded code

I think I got everything without breaking anything.

Also need a bit of help making the return statements in DisguiseOrderTargeter work better if possible. They work as is, just wondering if there's a better way to do them.

reaperrr commented:
Needs a rebase now that https://github.com/OpenRA/OpenRA/pull/13639 was merged (which also added `Damaged`, so `Move` would be the only new RevealDiguiseOn type).

I also think it would be better to squash/fixup all of your commits into one *before* you rebase.

pchote commented:
Any progress here @StormWingDelta?
